src/lib/hooks/swap/useSwapCallback.tsx

```diff
@@ -0,0 +1,175 @@
+          swapCalls.map((call) => {
+            const { address, calldata, value } = call
+
+            const tx =
+              !value || isZero(value)
+                ? { from: account, to: address, data: calldata }
+                : {
+                    from: account,
+                    to: address,
+                    data: calldata,
+                    value,
+                  }
+
+            return library
+              .estimateGas(tx)
+              .then((gasEstimate) => {
+                return {
+                  call,
+                  gasEstimate,
+                }
+              })
+              .catch((gasError) => {
+                console.debug('Gas estimate failed, trying eth_call to extract error', call)
+
+                return library
+                  .call(tx)
+                  .then((result) => {
+                    console.debug('Unexpected successful call after failed estimate gas', call, gasError, result)
+                    return { call, error: <Trans>Unexpected issue with estimating the gas. Please try again.</Trans> }
+                  })
+                  .catch((callError) => {
+                    console.debug('Call threw error', call, callError)
+                    return { call, error: swapErrorToUserReadableMessage(callError) }
+                  })
+              })
+          })
+        )
+
+        // a successful estimation is a bignumber gas estimate and the next call is also a bignumber gas estimate
+        let bestCallOption: SuccessfulCall | SwapCallEstimate | undefined = estimatedCalls.find(
+          (el, ix, list): el is SuccessfulCall =>
+            'gasEstimate' in el && (ix === list.length - 1 || 'gasEstimate' in list[ix + 1])
+        )
+
+        // check if any calls errored with a recognizable error
+        if (!bestCallOption) {
+          const errorCalls = estimatedCalls.filter((call): call is FailedCall => 'error' in call)
+          if (errorCalls.length > 0) throw errorCalls[errorCalls.length - 1].error
+          const firstNoErrorCall = estimatedCalls.find<SwapCallEstimate>(
+            (call): call is SwapCallEstimate => !('error' in call)
+          )
+          if (!firstNoErrorCall) throw new Error(t`Unexpected error. Could not estimate gas for the swap.`)
+          bestCallOption = firstNoErrorCall
+        }
+
+        const {
+          call: { address, calldata, value },
+        } = bestCallOption
+
+        return library
+          .getSigner()
+          .sendTransaction({
+            from: account,
+            to: address,
+            data: calldata,
+            // let the wallet try if we can't estimate the gas
+            ...('gasEstimate' in bestCallOption ? { gasLimit: calculateGasMargin(bestCallOption.gasEstimate) } : {}),
+            ...(value && !isZero(value) ? { value } : {}),
+          })
+          .then((response) => {
+            //@TODO(ianlapham): Use widget txn adding state to add transaction here
```

Comment:
Don't do this here - return the hash, but avoid coupling swap state to transactions state. Instead, have your component take the returned hash and add the transaction to transactions state. This way the component deals with state, and the states stay decoupled.

Reply:
yeah thats what i was intending here, add txn to our transactions state